src/ck-core/cklocation.h

```diff
@@ -481,8 +480,8 @@ class CkLocMgr : public IrrGroup
   IndexMsgBuffer bufferedIndexMsgs;
   IndexMsgBuffer bufferedDemandMsgs;
 
-  void deliverAnyBufferedMsgs(CmiUInt8, MsgBuffer& buffer);
-  void deliverAnyBufferedMsgs(const CkArrayIndex&, CmiUInt8, IndexMsgBuffer&);
+  void deliverAnyBufferedMsgs(CmiUInt8, MsgBuffer& buffer, bool firstAttempt = true);
```

Comment:
I'm worried that this also presents more issues. How do we know if the buffered messages have had an attempted delivery yet? What if some have, and some haven't? I think this just inherently will lead to potential inconsistencies. What would happen if instead, we record the send when the message is buffered? Even though the send is not carried out at that point, it is still the case that the program is attempting to communicate from elem X to elem Y at that point in execution...

Reply:
Although, I realize with the latter point I mentioned, we wouldn't be able to include the PE, but why is that even a necessary thing to include? Isn't the comm graph object to object?

Reply:
I agree that it's a bit of a flaky situation, the fact that `deliverMsg` is reused on both the send side and receive side makes it inherently a bit of a sticky wicket. I assigned the values for `firstAttempt` based on tracing through the call graph, making it `false` only for things that should have already gone through `deliverMsg` once before. Ideally, we'd have separate send-side and deliver-side functions for this so that we can record the send for LB purposes in a function that is clearly used only for sending a message once its destination and other details are known (as I believe your refactor of the location manager will do).

Yes, the comm graph is generally object to object, but it saves object locations for hops/hopbytes calculations that MetaBalancer uses as one of its decision criteria, so it needs object locations as well.